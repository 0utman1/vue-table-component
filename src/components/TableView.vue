<template>
    <div>
        <table>
            <thead>
            <tr>
                <th v-for="column in columns">
                    {{ column.name }}
                </th>
            </tr>
            </thead>
            <tbody>
                <table-row v-for="row in rows" :row="row" :columns="columns" />
            </tbody>
        </table>
    </div>
</template>

<script>
    import TableRow from './TableRow';

    export default {
        components: { TableRow },

        props: {
            rows: {
                required: true,
            },
            sortBy: {
                type: String,
                required: false,
                default: '',
            },
            sortOrder: {
                type: String,
                required: false,
                default: 'desc',
            }
        },

        data: () => ({}),

        created() {
            this.columns = this.$children;
        },

        computed: {},
    };

</script>
