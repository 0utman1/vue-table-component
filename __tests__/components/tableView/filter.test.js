import TableView from '../../../src';
import Vue from 'vue/dist/vue.js';

describe('Filterable tableView', () => {
    Vue.use(TableView);

    beforeEach(() => {
        document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-view
                        :data="[{ id: 1, firstName: 'Jay', lastName: 'Vleugels' },
                                { id: 2, firstName: 'Wesley', lastName: 'Biets' },
                                { id: 3, firstName: 'Randy', lastName: 'Paret' },
                                { id: 4, firstName: 'Devon', lastName: 'Macharis' }]"
                        sort-by="lastName"
                        sort-order="desc"
                    >
                        <table-column for="firstName" label="First name"></table-column>
                        <table-column for="lastName" label="Last name" :filterable="false"></table-column>
                    </table-view>
                </div>
            </div>
        `;


    });

    it('can filter data', async () => {

        const table = await createVm();

        table.filter = 'Wesley';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will filter data in a case-insensitive way', async () => {

        const table = await createVm();

        table.filter = 'wesley';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will display a message if there are no matching rows', async () => {

        const table = await createVm();

        table.filter = 'there are no rows that match this';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not use columns that are not filterable', async () => {

        const table = await createVm();

        table.filter = 'Vleugels';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });


});



async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});

    const table = vm.$children[0];

    return table ;
}
