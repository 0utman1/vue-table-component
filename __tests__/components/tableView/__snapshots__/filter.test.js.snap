// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filterable tableView can filter data 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <table><thead><tr><th>
    First name
</th><th class=\\"sort-desc\\">
    Last name
</th></tr></thead> <tbody><tr><td>Wesley</td><td>Biets</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableView will display a message if there are no matching rows 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <table><thead><tr><th>
    First name
</th><th class=\\"sort-desc\\">
    Last name
</th></tr></thead> <tbody></tbody></table> <div>
        There are no matching rows
    </div> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableView will filter data in a case-insensitive way 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <table><thead><tr><th>
    First name
</th><th class=\\"sort-desc\\">
    Last name
</th></tr></thead> <tbody><tr><td>Wesley</td><td>Biets</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableView will not use columns that are not filterable 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <table><thead><tr><th>
    First name
</th><th class=\\"sort-desc\\">
    Last name
</th></tr></thead> <tbody></tbody></table> <div>
        There are no matching rows
    </div> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;
